foolbox/attacks/newtonfool.py

```diff
@@ -0,0 +1,108 @@
+        """
+        Parameters
+        ----------
+        input_or_adv : `numpy.ndarray` or :class:`Adversarial`
+            The original, unperturbed input as a `numpy.ndarray` or
+            an :class:`Adversarial` instance.
+        label : int
+            The reference label of the original input. Must be passed
+            if `a` is a `numpy.ndarray`, must not be passed if `a` is
+            an :class:`Adversarial` instance.
+        unpack : bool
+            If true, returns the adversarial input, otherwise returns
+            the Adversarial object.
+        max_iter : int
+            The maximum number of iterations.
+        eta : float
+            the eta coefficient
+        """
+
+        a = input_or_adv
+        del input_or_adv
+        del label
+        del unpack
+
+        if not a.has_gradient():
+            return
+
+        if a.target_class() is not None:
+            logging.fatal('NewtonFool is an untargeted adversarial attack.')
+            return
+
+        reshape = np.reshape(a.original_image, [-1])
+        l2_norm = np.linalg.norm(reshape)
+        min_, max_ = a.bounds()
+        perturbed_image = a.original_image.copy()
+
+        for i in range(max_iter):
+
+            # (1) get the score and gradients
+            logits, gradients, is_adversarial = \
+                a.predictions_and_gradient(perturbed_image)
+
+            if is_adversarial:
+                return
+
+            score = np.max(softmax(logits))
+            # instead of using the logits and the gradient of the logits,
+            # we use a numerically stable implementation of the cross-entropy
+            # and expect that the deep learning frameworks also use such a
+            # stable implemenation to calculate the gradient
+            # grad is calculated from CE but we want softmax
+            # -> revert chain rule
+            gradients = - gradients / score
+
+            # (2) calculate gradient norm
+            gradient_l2_norm = np.linalg.norm(np.reshape(gradients, [-1]))
+
+            # (3) calculate delta
+            delta = self._delta(eta, l2_norm, score,
+                                gradient_l2_norm, a.num_classes())
+
+            # delta = 0.01
+
+            # (4) calculate & apply current pertubation
+            current_pertubation = self._perturbation(delta,
+                                                     gradients,
+                                                     gradient_l2_norm)
+
+            perturbed_image += current_pertubation
+            perturbed_image = np.clip(perturbed_image, min_, max_)
+
+    @staticmethod
+    def _delta(eta, norm, score, gradient_norm, num_classes):
+        a = eta * norm * gradient_norm
+        b = score - 1.0 / num_classes
+        return min(a, b)
+
+    @staticmethod
+    def _perturbation(delta, gradients, gradient_norm):
+        a = delta * gradients
+        b = float(gradient_norm ** 2)
+        direction = -(a / b)
```

Comment:
nitpick: unless there are numerical reasons, you could safe one matrix-scalar product by first calculating (delta / b) and then multiplying with gradients